builder/builder.go

```diff
@@ -351,148 +426,390 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
 	return errors.Wrapf(os.Remove(tmptarfile), "Failed to remove temp bundle archive: %s", tmptarfile)
 }
 
-// AddBundles will copy the specified upstream bundles from the configured
-// upstream version to the mix-bundles directory. In error event, the int value
-// returned will contain the number of bundles added thus far.
-// bundles: array slice of bundle names
-// force: override bundle in mix-dir when present
-// all: include all CLR bundles. Overrides bundles.
-// git: automatically git commit with bundles added
-func (b *Builder) AddBundles(bundles []string, force bool, allbundles bool, git bool) (int, error) {
-	var bundleAddCount int
+// getBundlePath returns the path to the bundle definition file for a given
+// bundle name, or error if it cannot be found. Looks first in local-bundles,
+// then upstream-bundles.
+func (b *Builder) getBundlePath(bundle string) (string, error) {
+	// Check local-bundles
+	path := filepath.Join(b.LocalBundleDir, bundle)
+	if _, err := os.Stat(path); err == nil {
+		return path, nil
+	}
 
-	// Check if mix bundles dir exists
-	if _, err := os.Stat(b.Bundledir); os.IsNotExist(err) {
-		return bundleAddCount, errors.New("Mix bundles directory does not exist. Run mixer init-mix")
+	// Check upstream-bundles
+	path = filepath.Join(getUpstreamBundlesPath(b.UpstreamVer), bundle)
+	if _, err := os.Stat(path); err == nil {
+		return path, nil
 	}
 
-	clrbundledir := getUpstreamBundlesPath(b.Clearver)
+	return "", errors.Errorf("Cannot find bundle %s in local or upstream bundles", bundle)
+}
+
+// isLocalBundle checks a bundle filepath is local
+func (b *Builder) isLocalBundle(path string) bool {
+	return strings.HasPrefix(path, b.LocalBundleDir)
+}
+
+type stringSet map[string]bool
 
-	// Fetch upstream bundles if needed
-	if err := b.getUpstreamBundles(b.Clearver, true); err != nil {
-		return bundleAddCount, err
+func getStringSetKeys(set stringSet) []string {
+	keys := make([]string, len(set))
+	i := 0
+	for k := range set {
+		keys[i] = k
+		i++
 	}
+	return keys
+}
 
-	// Add all bundles to list if --all passed
-	if allbundles {
-		files, err := ioutil.ReadDir(clrbundledir)
-		if err != nil {
-			return bundleAddCount, err
-		}
+func getStringSetKeysSorted(set stringSet) []string {
+	keys := getStringSetKeys(set)
+	sort.Strings(keys)
+	return keys
+}
+
+func getBundleSetKeysSorted(set bundleSet) []string {
+	keys := make([]string, len(set))
+	i := 0
+	for k := range set {
+		keys[i] = k
+		i++
+	}
```

Comment:
Yup, I _hated_ writing this function twice. Spent a good half hour just trying to get rid of this stupid thing. It's the product of :
1. Go not supporting generics, 
1. My inability to get the pseudo-generic `getSetKeysSorted(set interface{})` to work with the custom `stringSet` and `bundleSet` types. (It's entirely possible I just don't know what I was doing wrong, and that approach might work... Though it would push the burden of casting to and from a type to the caller...), and 
1. A bunch of golang blogs telling me to just duplicate it for something this basic.

But I'm strongly considering your "just use bundleSet everywhere and ignore the overhead" suggestion below, in which case this duplication would disappear.

Reply:
Generics ftw, maybe next Go release :+1:  If you/anyone else doesn't have any concern about using bundleSet everywhere it seems easier, since most of these commands are one-shot type deals. I don't think interfaces{} are necessarily better either.